### PR TITLE
Bug 208680 - Support read only filesystems and load params from db

### PR DESCRIPTION
**Pull request by dylanwh**

### CODE REVIEW

**Comment by dklawren** [comment]
t/011pod.t ........... 205/379
#   Failed test 'Bugzilla/Config.pm POD coverage is 75%. Undocumented methods: write_param_file, read_params'
#   at t/011pod.t line 109.
t/011pod.t ........... 351/379 # Looks like you failed 1 test of 379.
t/011pod.t ........... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/379 subtests

Also why not just forget about read_only_fs and place the params in the DB under normal circumstances? This helps for multiple web heads to share the same params without using a shared filesystem. 

Also this does not address the issue with being able to show a shutdown message if the DB is inaccessible.

When trying to save the params with read_only_fs=1, I get the following taint error:

**Comment by dklawren** [comment]
Failed to write params to database!: Insecure dependency in parameter 2 of DBI::st=HASH(0x531fc30)->execute method call while running with -T switch at Bugzilla/Config.pm line 306.

**Comment by dylanwh** [comment]
oh snap, I never tested under taint mode because (currently) it is impossible to run under taint mode with plack.

**Comment by dklawren** [comment]
Adding new table params...
Use of uninitialized value $_ in hash element at Bugzilla/Config.pm line 309.
Removing existing compiled templates...
Precompiling templates...done.
Fixing file permissions...

Running checksetup.pl for first time on an already initialized DB after applying patch.

File: Bugzilla/Config.pm [hunk 1]
```diff
@@ -307,25 +307,26 @@ sub write_params {
         my $ok = eval {
             $dbh->do("DELETE FROM params");
             my $sth = $dbh->prepare("INSERT INTO params (param_key, param_val) VALUES (?, ?)");
+            trick_taint($params->{$_}) if defined $params->{$_};
             $sth->execute($_, $params->{$_}) foreach keys %$params;
             $dbh->bz_commit_transaction();
             1;
         };
```
Comment (dklawren) [inline Bugzilla/Config.pm]: This does not work. Gives taint error still. Needs to be instead:

     my $ok = eval {
        $dbh->do("DELETE FROM params");
        my $sth = $dbh->prepare("INSERT INTO params (param_key, param_val) VALUES (?, ?)");
        foreach my $key (keys %$params) {
          trick_taint($params->{$key}) if defined $params->{$key};
          $sth->execute($key, $params->{$key});
      }
      $dbh->bz_commit_transaction();
      1;
   };

**Pull request closed without merging**